File: backend/core-api/src/test/java/ai/govbiz/core/client/ai/AiServiceClientTest.java
package ai.govbiz.core.client.ai;

import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.net.http.HttpConnectTimeoutException;
import java.util.List;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.test.web.client.MockRestServiceServer;
import org.springframework.web.client.RestClient;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.header;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.method;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withException;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withNoContent;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withStatus;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;

class AiServiceClientTest {

    private static final String BASE_URL = "http://ai-service.test:8000";
    private static final String HEALTH_URL = BASE_URL + "/internal/v1/health";
    private static final String SEARCH_INTENT_URL =
            BASE_URL + "/internal/v1/search-intents/analyze";
    private static final String VALID_RESPONSE = """
            {"status":"up","service":"govbiz-ai-service"}
            """;
    private static final String VALID_INTENT_RESPONSE = """
            {
              "originalQuery":"서울 AI 스타트업 지원사업",
              "keywords":["스타트업"],
              "regions":["서울"],
              "categories":["AI","창업"],
              "targetTerms":["창업기업"],
              "acceptingOnly":true,
              "clarificationNeeded":false,
              "clarificationQuestion":null
            }
            """;

    private MockRestServiceServer server;
    private AiServiceClient client;

    @BeforeEach
    void setUp() {
        RestClient.Builder builder = RestClient.builder()
                .baseUrl(BASE_URL)
                .defaultHeader(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE);
        server = MockRestServiceServer.bindTo(builder).build();
        client = new AiServiceClient(builder.build());
    }

    @AfterEach
    void verifiesEveryExpectedRequest() {
        server.verify();
    }

    @Test
    void sendsExactHealthRequestAndDecodesValidJson() {
        server.expect(requestTo(HEALTH_URL))
                .andExpect(method(HttpMethod.GET))
                .andExpect(header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE))
                .andRespond(withSuccess(VALID_RESPONSE, MediaType.APPLICATION_JSON));

        AiServiceHealthPayload response = client.getHealth();

        assertEquals("up", response.status());
        assertEquals("govbiz-ai-service", response.service());
    }

    @Test
    void mapsDownstream4xxToUpstreamError() {
        expectResponse(withStatus(HttpStatus.NOT_FOUND)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"));

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR);
    }

    @Test
    void mapsDownstream5xxToUpstreamError() {
        expectResponse(withStatus(HttpStatus.SERVICE_UNAVAILABLE)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"));

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR);
    }

    @Test
    void mapsUnexpectedSuccessfulStatusToUpstreamError() {
        expectResponse(withStatus(HttpStatus.CREATED)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"));

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR);
    }

    @Test
    void mapsRedirectStatusToUpstreamErrorWithoutDecodingBody() {
        expectResponse(withStatus(HttpStatus.FOUND)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"));

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR);
    }

    @Test
    void mapsNoContentToInvalidResponse() {
        expectResponse(withNoContent());

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE);
    }

    @Test
    void mapsEmptySuccessBodyToInvalidResponse() {
        expectResponse(withSuccess("", MediaType.APPLICATION_JSON));

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE);
    }

    @Test
    void mapsWrongContentTypeToInvalidResponse() {
        expectResponse(withSuccess(VALID_RESPONSE, MediaType.TEXT_PLAIN));

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE);
    }

    @Test
    void mapsMalformedJsonToInvalidResponse() {
        expectResponse(withSuccess("{\"status\":", MediaType.APPLICATION_JSON));

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE);
    }

    @Test
    void mapsConnectionFailureToUnavailable() {
        expectResponse(withException(new ConnectException("connection refused")));

        assertFailure(AiServiceClientException.Failure.UNAVAILABLE);
    }

    @Test
    void mapsConnectTimeoutToTimeout() {
        expectResponse(withException(new HttpConnectTimeoutException("connect timeout")));

        assertFailure(AiServiceClientException.Failure.TIMEOUT);
    }

    @Test
    void mapsReadTimeoutToTimeout() {
        expectResponse(withException(new SocketTimeoutException("read timeout")));

        assertFailure(AiServiceClientException.Failure.TIMEOUT);
    }

    @Test
    void sendsExactSearchIntentRequestAndDecodesTheStructuredResponse() {
        server.expect(requestTo(SEARCH_INTENT_URL))
                .andExpect(method(HttpMethod.POST))
                .andExpect(header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE))
                .andExpect(header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
                .andExpect(org.springframework.test.web.client.match.MockRestRequestMatchers.content()
                        .json("""
                                {
                                  "query":"서울 AI 스타트업 지원사업",
                                  "acceptingOnly":true
                                }
                                """))
                .andRespond(withSuccess(VALID_INTENT_RESPONSE, MediaType.APPLICATION_JSON));

        AiSearchIntentPayload response = client.analyzeSearchIntent(
                "서울 AI 스타트업 지원사업",
                true);

        assertEquals("서울 AI 스타트업 지원사업", response.originalQuery());
        assertEquals(List.of("서울"), response.regions());
        assertEquals(List.of("AI", "창업"), response.categories());
    }

    @Test
    void mapsSearchIntentNoContentToInvalidResponse() {
        server.expect(requestTo(SEARCH_INTENT_URL)).andRespond(withNoContent());

        assertIntentFailure(AiServiceClientException.Failure.INVALID_RESPONSE);
    }

    @Test
    void mapsMalformedSearchIntentJsonToInvalidResponse() {
        server.expect(requestTo(SEARCH_INTENT_URL))
                .andRespond(withSuccess("{\"keywords\":", MediaType.APPLICATION_JSON));

        assertIntentFailure(AiServiceClientException.Failure.INVALID_RESPONSE);
    }

    @Test
    void mapsSearchIntentUnavailableStatusToUnavailable() {
        server.expect(requestTo(SEARCH_INTENT_URL))
                .andRespond(withStatus(HttpStatus.SERVICE_UNAVAILABLE)
                        .contentType(MediaType.APPLICATION_JSON)
                        .body("{\"detail\":\"unavailable\"}"));

        assertIntentFailure(AiServiceClientException.Failure.UNAVAILABLE);
    }

    @Test
    void mapsSearchIntentGatewayTimeoutStatusToTimeout() {
        server.expect(requestTo(SEARCH_INTENT_URL))
                .andRespond(withStatus(HttpStatus.GATEWAY_TIMEOUT)
                        .contentType(MediaType.APPLICATION_JSON)
                        .body("{\"detail\":\"timeout\"}"));

        assertIntentFailure(AiServiceClientException.Failure.TIMEOUT);
    }

    private void expectResponse(org.springframework.test.web.client.ResponseCreator response) {
        server.expect(requestTo(HEALTH_URL)).andRespond(response);
    }

    private void assertFailure(AiServiceClientException.Failure expectedFailure) {
        AiServiceClientException exception = assertThrows(
                AiServiceClientException.class,
                client::getHealth);
        assertEquals(expectedFailure, exception.failure());
    }

    private void assertIntentFailure(AiServiceClientException.Failure expectedFailure) {
        AiServiceClientException exception = assertThrows(
                AiServiceClientException.class,
                () -> client.analyzeSearchIntent("서울 AI", true));
        assertEquals(expectedFailure, exception.failure());
    }
}

File: backend/core-api/src/test/kotlin/ai/govbiz/core/client/ai/AiServiceClientTest.kt
package ai.govbiz.core.client.ai

import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.http.HttpConnectTimeoutException
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.test.web.client.MockRestServiceServer
import org.springframework.test.web.client.ResponseCreator
import org.springframework.test.web.client.match.MockRestRequestMatchers.content
import org.springframework.test.web.client.match.MockRestRequestMatchers.header
import org.springframework.test.web.client.match.MockRestRequestMatchers.method
import org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo
import org.springframework.test.web.client.response.MockRestResponseCreators.withException
import org.springframework.test.web.client.response.MockRestResponseCreators.withNoContent
import org.springframework.test.web.client.response.MockRestResponseCreators.withStatus
import org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess
import org.springframework.web.client.RestClient

class AiServiceClientTest {

    private lateinit var server: MockRestServiceServer
    private lateinit var client: AiServiceClient

    @BeforeEach
    fun setUp() {
        val builder = RestClient.builder()
            .baseUrl(BASE_URL)
            .defaultHeader(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
        server = MockRestServiceServer.bindTo(builder).build()
        client = AiServiceClient(builder.build())
    }

    @AfterEach
    fun verifiesEveryExpectedRequest() {
        server.verify()
    }

    @Test
    fun sendsExactHealthRequestAndDecodesValidJson() {
        server.expect(requestTo(HEALTH_URL))
            .andExpect(method(HttpMethod.GET))
            .andExpect(header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE))
            .andRespond(withSuccess(VALID_RESPONSE, MediaType.APPLICATION_JSON))

        val response = client.getHealth()

        assertEquals("up", response.status)
        assertEquals("govbiz-ai-service", response.service)
    }

    @Test
    fun mapsDownstream4xxToUpstreamError() {
        expectResponse(
            withStatus(HttpStatus.NOT_FOUND)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"),
        )

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR)
    }

    @Test
    fun mapsDownstream5xxToUpstreamError() {
        expectResponse(
            withStatus(HttpStatus.SERVICE_UNAVAILABLE)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"),
        )

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR)
    }

    @Test
    fun mapsUnexpectedSuccessfulStatusToUpstreamError() {
        expectResponse(
            withStatus(HttpStatus.CREATED)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"),
        )

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR)
    }

    @Test
    fun mapsRedirectStatusToUpstreamErrorWithoutDecodingBody() {
        expectResponse(
            withStatus(HttpStatus.FOUND)
                .contentType(MediaType.TEXT_PLAIN)
                .body("not-json"),
        )

        assertFailure(AiServiceClientException.Failure.UPSTREAM_ERROR)
    }

    @Test
    fun mapsNoContentToInvalidResponse() {
        expectResponse(withNoContent())

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE)
    }

    @Test
    fun mapsEmptySuccessBodyToInvalidResponse() {
        expectResponse(withSuccess("", MediaType.APPLICATION_JSON))

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE)
    }

    @Test
    fun mapsWrongContentTypeToInvalidResponse() {
        expectResponse(withSuccess(VALID_RESPONSE, MediaType.TEXT_PLAIN))

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE)
    }

    @Test
    fun mapsMalformedJsonToInvalidResponse() {
        expectResponse(withSuccess("{\"status\":", MediaType.APPLICATION_JSON))

        assertFailure(AiServiceClientException.Failure.INVALID_RESPONSE)
    }

    @Test
    fun mapsConnectionFailureToUnavailable() {
        expectResponse(withException(ConnectException("connection refused")))

        assertFailure(AiServiceClientException.Failure.UNAVAILABLE)
    }

    @Test
    fun mapsConnectTimeoutToTimeout() {
        expectResponse(withException(HttpConnectTimeoutException("connect timeout")))

        assertFailure(AiServiceClientException.Failure.TIMEOUT)
    }

    @Test
    fun mapsReadTimeoutToTimeout() {
        expectResponse(withException(SocketTimeoutException("read timeout")))

        assertFailure(AiServiceClientException.Failure.TIMEOUT)
    }

    @Test
    fun sendsExactRankingRequestAndDecodesTheStructuredScores() {
        server.expect(requestTo(RANKING_URL))
            .andExpect(method(HttpMethod.POST))
            .andExpect(header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE))
            .andExpect(header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE))
            .andExpect(
                content().json(
                    """
                    {
                      "originalQuery":"서울 AI 스타트업 지원사업",
                      "scoringVersion":"govbiz-support-program-ranking-v1",
                      "resultLimit":1,
                      "candidates":[{
                        "id":"program-1",
                        "title":"서울 AI 사업화 지원",
                        "organization":"서울경제진흥원",
                        "summary":"AI 창업기업 사업화 지원",
                        "categories":["AI","창업"],
                        "regions":["서울"],
                        "targetDescription":"서울 창업기업",
                        "applicationPeriod":"상시 접수",
                        "status":"OPEN"
                      }]
                    }
                    """.trimIndent(),
                ),
            )
            .andRespond(withSuccess(VALID_RANKING_RESPONSE, MediaType.APPLICATION_JSON))

        val response = client.rankSupportPrograms(rankingRequest())

        assertEquals("서울 AI 스타트업 지원사업", response.originalQuery)
        assertEquals(95, response.rankings?.single()?.totalScore)
        assertEquals("program-1", response.rankings?.single()?.programId)
    }

    @Test
    fun mapsRankingNoContentToInvalidResponse() {
        server.expect(requestTo(RANKING_URL)).andRespond(withNoContent())

        assertRankingFailure(AiServiceClientException.Failure.INVALID_RESPONSE)
    }

    @Test
    fun mapsMalformedRankingJsonToInvalidResponse() {
        server.expect(requestTo(RANKING_URL))
            .andRespond(withSuccess("{\"rankings\":", MediaType.APPLICATION_JSON))

        assertRankingFailure(AiServiceClientException.Failure.INVALID_RESPONSE)
    }

    @Test
    fun mapsRankingUnavailableStatusToUnavailable() {
        server.expect(requestTo(RANKING_URL))
            .andRespond(
                withStatus(HttpStatus.SERVICE_UNAVAILABLE)
                    .contentType(MediaType.APPLICATION_JSON)
                    .body("{\"detail\":\"unavailable\"}"),
            )

        assertRankingFailure(AiServiceClientException.Failure.UNAVAILABLE)
    }

    @Test
    fun mapsRankingGatewayTimeoutStatusToTimeout() {
        server.expect(requestTo(RANKING_URL))
            .andRespond(
                withStatus(HttpStatus.GATEWAY_TIMEOUT)
                    .contentType(MediaType.APPLICATION_JSON)
                    .body("{\"detail\":\"timeout\"}"),
            )

        assertRankingFailure(AiServiceClientException.Failure.TIMEOUT)
    }

    private fun expectResponse(response: ResponseCreator) {
        server.expect(requestTo(HEALTH_URL)).andRespond(response)
    }

    private fun assertFailure(expectedFailure: AiServiceClientException.Failure) {
        val exception = assertThrows(AiServiceClientException::class.java) {
            client.getHealth()
        }
        assertEquals(expectedFailure, exception.failure)
    }

    private fun assertRankingFailure(expectedFailure: AiServiceClientException.Failure) {
        val exception = assertThrows(AiServiceClientException::class.java) {
            client.rankSupportPrograms(rankingRequest())
        }
        assertEquals(expectedFailure, exception.failure)
    }

    private fun rankingRequest() = AiSupportProgramRankingRequest(
        originalQuery = "서울 AI 스타트업 지원사업",
        scoringVersion = "govbiz-support-program-ranking-v1",
        resultLimit = 1,
        candidates = listOf(
            AiSupportProgramCandidateRequest(
                id = "program-1",
                title = "서울 AI 사업화 지원",
                organization = "서울경제진흥원",
                summary = "AI 창업기업 사업화 지원",
                categories = listOf("AI", "창업"),
                regions = listOf("서울"),
                targetDescription = "서울 창업기업",
                applicationPeriod = "상시 접수",
                status = "OPEN",
            ),
        ),
    )

    private companion object {
        const val BASE_URL = "http://ai-service.test:8000"
        const val HEALTH_URL = "$BASE_URL/internal/v1/health"
        const val RANKING_URL = "$BASE_URL/internal/v1/support-program-rankings/rank"
        val VALID_RESPONSE =
            """
            {"status":"up","service":"govbiz-ai-service"}
            """.trimIndent()
        val VALID_RANKING_RESPONSE =
            """
            {
              "originalQuery":"서울 AI 스타트업 지원사업",
              "scoringVersion":"govbiz-support-program-ranking-v1",
              "rankings":[{
                "programId":"program-1",
                "semanticRelevance":38,
                "targetFit":24,
                "regionFit":15,
                "applicationStatusFit":10,
                "supportTypeFit":8,
                "totalScore":95,
                "recommendationReasons":["서울 AI 창업기업 사업화 지원"]
              }]
            }
            """.trimIndent()
    }
}
